### PR TITLE
Better login via load functions

### PR DESCRIPTION
**Pull request by rmunn**
The preferred approach for keeping track of users in Svelte-Kit is as follows:

- The `handle` hook adds a `getUser` function to `event.locals`, which will be called later on when user info is needed
    - The function closes over `event.cookies`, so it can be called later on without parameters
- The root layout, in the **server** `load` function, calls `getUser` and puts user info into its return value
- Any load function that needs user data can do `const { user } = await event.parent()`
- Any *page* that needs user data can access `$page.data.user`
- Most important: if someone signs out, the load function re-runs itself so that `$page.data.user` is always up-to-date and doesn't end up getting stale

This PR modifies the login logic to follow that algorithm.

### CODE REVIEW

**Comment by rmunn** [comment]
I noticed eslint was warning about unsafe types in the AdminContent.svelte page, because I used `$page.data.user` which it thought had an `any` type. I did a quick fix by silencing eslint about that line, but according to https://kit.svelte.dev/docs/types#app-pagedata we should actually edit `app.d.ts` and add a `PageData` interface containing the items that we are guaranteeing to be available in the page data, such as the user. So I'll come back to this at some point and add proper typing.

File: frontend/src/lib/user.ts
Comment (myieye) [inline frontend/src/lib/user.ts:119]:
```suggestion
```

File: frontend/src/lib/user.ts
Comment (myieye) [inline frontend/src/lib/user.ts:134]: Shouldn't we just throw the error again?
If we ever land in here we want to know about it right?

File: frontend/src/lib/layout/AppMenu.svelte
Comment (myieye) [inline frontend/src/lib/layout/AppMenu.svelte:14]: Ultimately we could create a user store that derives from `$page` right?
We can configure lint to forbid certain imports in certain places. I'm looking forward to have a good reason to make use of that feature 😆

**Review by myieye — CHANGES_REQUESTED** [review]
Just a few thoughts and a comment that wants to be removed 🙂

File: frontend/src/lib/user.ts
Comment (rmunn) [reply frontend/src/lib/user.ts:119]: I'm actually thinking I'll put this line in and remove the more elaborate try...catch block, because if `import('node:crypto')` fails then it will throw an exception, which is just what you were suggesting in the other comment.

BTW, the reason for this is because we were running into `https://github.com/nodejs/node/issues/48444` during our docker build process, so I had to temporarily downgrade the Node version to 18 instead of 20 to work around that bug — and Node 18 doesn't have the same crypto modules. This workaround will be able to go away soon, though: that issue is affecting a lot of people so they're fast-tracking a new release of the affected library. At which point we'll be able to go back to building with modern Node versions and this hack can be removed.

File: frontend/src/lib/user.ts
Comment (rmunn) [reply frontend/src/lib/user.ts:134]: As mentioned above, I'm going to make a change that will end up throwing the exception. I put this part in, swallowing the exception, so that I could actually get `docker build` to succeed, otherwise I was being blocked on my currrent work.

File: frontend/src/lib/layout/AppMenu.svelte
Comment (rmunn) [reply frontend/src/lib/layout/AppMenu.svelte:14]: We could, but I don't think it's a good idea. `$page.data.user` is type-safe, and doesn't break the Law of Demeter because `$page.data` is guaranteed by the Svelte docs to always exist. So `user = derived(page, p => p.data.user)` is just unnecessary complexity IMHO.

**Comment by rmunn** [comment]
Interesting. The compilation is failing on `src/routes/(authenticated)/myaccount/+page.svelte`, which imports the user store that I'm getting rid of. Thing is, I don't have that file in my repo at all, so I didn't see it: it was added to `develop` later than when my PR was created. Which means that what GitHub Actions is building is the result of an attempted merge between the PR branch and develop. **Which is exactly what we want.** It's telling me about problems that *would* arise if I were to merge this, which I could not have known about short of merging `develop` into my branch on a daily basis (which is, usually, something you want to avoid).

Now, here, I *do* need to merge develop into my branch and then fix the resulting issues. But I get to do it only on an as-needed basis because GHA told me that it needs to happen. This just goes to show the advantage of continuous integration.

**Comment by myieye** [comment]
@rmunn FYI I'm going to try to fix the GitHub actions, so we don't have duplicate runs. I turned on the pull_request actions so that they run against @dadukhankevin's fork PRs.

File: frontend/src/lib/otel/shared.ts
Comment (rmunn) [reply frontend/src/lib/otel/shared.ts:87]: Had to do it this way because of the way the Typescript checks work, but I can write different checks. (`ifBrowserEvent`, if it returns false, removes `Event` from the list of possibilities, at which point we are able to call `isNavigationEvent`. But I can change the types of the functions to achieve the same result.

File: frontend/src/lib/otel/shared.ts
Comment (rmunn) [reply frontend/src/lib/otel/shared.ts:107]: Yes, you're right. I originally meant to have getUser called inside the else clause, but the way I wrote it, it can now be moved outside the if and we'll be fine.

File: frontend/src/lib/otel/shared.ts
Comment (myieye) [reply frontend/src/lib/otel/shared.ts:87]: Oh, gotcha! Make isNavigationEvent should be more forgiving/also accept Events.

Comment (rmunn) [reply frontend/src/lib/otel/shared.ts]: Tweaked the Typescript types that `isRequestEvent` accepts just slightly, and thereby enabled myself to merge the two else clauses together. Now I'm much happier with it, and the other use of `isRequestEvent` is not affected.

File: frontend/src/lib/otel/shared.ts
Comment (rmunn) [reply frontend/src/lib/otel/shared.ts]: Got rid of it, and also the `getUserId` function from lib/user.ts which nothing is now calling.

**Comment by rmunn** [comment]
> @rmunn FYI I'm going to try to fix the GitHub actions, so we don't have duplicate runs. I turned on the pull_request actions so that they run against @dadukhankevin's fork PRs.

I suggest triggering the `push` build only on pushes to `develop` or `master`, then the `pull_request` trigger will take care of other push situations. That would mean that the actions would *not* run on a push to a branch that one hasn't opened a PR on, but I think that's what we want.

**Comment by rmunn** [comment]
> Also FYI, you can run `pnpm run lint` locally :wink:

I'm just going to install the eslint VS Code plugin so I get red squigglies in the editor, that will be the best way for me to remember it.

**Comment by myieye** [comment]
> > @rmunn FYI I'm going to try to fix the GitHub actions, so we don't have duplicate runs. I turned on the pull_request actions so that they run against @dadukhankevin's fork PRs.
> 
> I suggest triggering the `push` build only on pushes to `develop` or `master`, then the `pull_request` trigger will take care of other push situations. That would mean that the actions would _not_ run on a push to a branch that one hasn't opened a PR on, but I think that's what we want.

Yeah, that's pretty much what [I just did](https://github.com/sillsdev/languageforge-lexbox/blob/cb8089a551b7064a86fd4d73d967279f65af8132/.github/workflows/lexbox-ui.yaml#L4).
I'm not quite sure if it's doing what it should be 🤔, because the "push" got triggered here again.

**Comment by rmunn** [comment]
> Yeah, that's pretty much what [I just did](https://github.com/sillsdev/languageforge-lexbox/blob/cb8089a551b7064a86fd4d73d967279f65af8132/.github/workflows/lexbox-ui.yaml#L4). I'm not quite sure if it's doing what it should be thinking, because the "push" got triggered here again.

Actions based on pull requests run using the workflow file from the PR branch, so you won't see the result of the change until develop is merged into this branch, or until we merge the PR. Which I think is ready to merge, so I won't bother pulling develop back in because I don't expect any more commits here.

File: frontend/src/routes/+layout.server.ts
Comment (myieye) [inline frontend/src/routes/+layout.server.ts:8]: But invalidating still won't have any effect until we implement the cookie getting updated right?

**Review by myieye — APPROVED** [review]

File: frontend/src/routes/+layout.server.ts
Comment (rmunn) [reply frontend/src/routes/+layout.server.ts:8]: Yes, at the moment this will just re-read the cookie. But once the `getUser` function is changed to actually load the user data from GraphQL, then the invalidate will start doing something.

**Pull request merged**
